Feat: Canvas Clipboard

cqliu1 opened the pull request:
Phase 1 of #25543.
~~Blocked by #25935.~~ merged

This implements simple copy/cut/paste for elements only and stores the copied elements in `localStorage`. 

![nov-19-2018 14-22-05](https://user-images.githubusercontent.com/1697105/48735882-d1494900-ec06-11e8-9261-1dc24774f2f7.gif)

By storing the copied elements in localStorage, the clipboard contents persist until cleared by the user and allows you to paste elements between pages of a workpad and between workpads.  

Key Mappings:
Copy: `alt` + `c`
Cut: `alt` + `x`
Paste: `alt` + `v`

Changes:
- adds copy, cut, and paste keyboard shortcuts for elements
- adds constants for Canvas localStorage keys 
- triggers toast when copying or cutting an element to confirm element was saved to clipboard

elasticmachine commented:
Pinging @elastic/kibana-canvas

cqliu1 commented:
@monfera I'm getting some odd behavior when trying to reposition pasted elements.

![nov-19-2018 14-26-26](https://user-images.githubusercontent.com/1697105/48736006-30a75900-ec07-11e8-9144-f290896706b1.gif)

I looked at the redux state while dragging around the pasted elements, and it looks like the position isn't being updated in redux.

Steps to reproduce:
1. Create any element
2. Copy element (`alt` + `c`)
3. Paste element (`alt` + `v`)
4. Drag pasted element to a new position
5. Paste element again (`alt` + `v`)
6. Observe the pasted elements all snap back to the original position

I'm thinking the position is possibly being updated in the aeroelastic store but not redux. Can you take a look at this PR when you get a chance? I'm not sure where the discrepancy is.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/591/)

cqliu1 commented:
@monfera thanks for fixing the positioning bug!

monfera commented:
Just to document, [this PR](https://github.com/elastic/kibana/pull/25935) fixes the glitch reported in https://github.com/elastic/kibana/pull/25890#issuecomment-440048504

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/705/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/853/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/905/)

cqliu1 commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/911/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/924/)

cqliu1 commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/952/)

cqliu1 commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/988/)

monfera commented:
retest

monfera commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1007/)

monfera commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1019/)

monfera commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1031/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1039/)

monfera commented:
There are often intermittent failures but there's a pattern, [this](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1039/JOB=x-pack-ciGroup2,node=immutable/) fails every time

alexfrancoeur commented:
Having a Mac, I attempted to use `cmd+c`, `cmd+x` and `cmd+v` for this functionality. That didn't work, so I went to the Windows keyboard shortcuts of `ctrl+c`, `ctrl+x` and `ctrl+v`. This also did not work, so I had to read the description of the issue in order to see what the keyboard shortcut was. I attempted to use `option` / `alt` with my bluetooth keyboard and my macbook keyboard but I can't seem to get it working. I've tried clearing the application cache as well and am using the most recent version of master with this PR. Is anyone else having this issue? I tried with a new workpad as well as an older one.

In general, I'm skeptical to use `alt` as it is not as discoverable as the commonly used keyboard shortcuts for this functionality. Is there any reason why we cannot use the more common combinations for each OS? While this could certainly be a learned behavior, it does not make for the best initial experience with Canvas. I imagine most users won't attempt to read the documentation in order to copy and paste elements. 

Do we need to have another button for copy or should we just replace the clone button functionality with copy functionality? @ryankeairns may be able to provide some insight.

Happy to provide more detail once I'm able to get the clipboard working.

monfera commented:
indeed `alt-c alt-v` are not discoverable, then later, switching keystrokes to the standard ones would be another annoyance, maybe [this approximation](https://github.com/elastic/kibana/pull/25935/files#diff-4439e21e439b060461f0ba2ec0aced56) could be extracted (see `appleKeyboard` and the gh issue Conversation

ryankeairns commented:
I've read through the linked issue, but am not following the relationship... I'll reiterate Alex's question, are we not able to use the standard command (control on Windows) keys for copy/cut/paste? I feel like there is a reason why not and that I have missed the point.

Thanks for clarifying!

monfera commented:
I'm not aware of a fundamental difficulty (Ryan, my link shows that we handle all the [various modifier keys](https://github.com/elastic/kibana/pull/25935#issue-232333320) elsewhere), I think it was simply for expediency to suggest the merge of the PR as soon as possible with whatever keyboard strokes. I'm on mandatory support training this week again, but maybe I can sneak in some time toward adding a commit tomorrow, unless there's some deeper obstacle @cqliu1?

cqliu1 commented:
I chose to map the shortcuts to the `alt-c` and `alt-v` so I could get the functionality implemented without running into to conflicts with the default behavior of `ctrl-c` and `ctrl-v` on PC and `cmd-c` and `cmd-v` on Mac. 

It also looked like all our shortcuts were using `alt` instead of `ctrl`/`cmd`. I can go ahead and try switching it over to ctrl/cmd and see if I run into any issues.

alexfrancoeur commented:
Thanks @cqliu1 please let me know once you've had a chance to switch to ctrl/cmd. I think this will be important for the feature. I'll take the PR for a spin again once this has been updated. I think I know why I wasn't able to get the clipboard functionality working in the first place. I'll provide additional feedback once you've made an attempt to using the different shortcuts.

cqliu1 commented:
@alexfrancoeur I updated the key mappings to use ctrl/cmd for cut/copy/paste.  I also updated the existing shortcuts to OS specific mappings. This is ready for a test run when you've got time.

alexfrancoeur commented:
Woot! Thank you @cqliu1, this is great! I don't have a windows machine so can't confirm the OS key mappings work there, but it does on mac.

The only issue I ran into was copy and pasting an element that has an image asset into a new workpad. Is it possible to copy and paste the asset over with the same ID as well?

![image](https://user-images.githubusercontent.com/25181823/49244857-c9e02780-f3de-11e8-8f7e-a42695ef6c99.png)

Everything else seems to work great. There were two other observations I had but I don't consider them blockers.

The clipboard seems to only be persisted within Kibana (though I haven't tried Kibana to Kibana). Copy, cut and paste work between pages and workpads. It works if you leave the Canvas app and come back as well. I expected to be able to paste outside of Kibana as well but that doesn't seem to work. I don't necessarily see this as in scope of this PR. If someone wants to copy and paste the expression into a text editor, they can go into the expression editor.

I noticed that if you copy and paste a group of elements that group is not persisted when you paste. See the GIF below. Again, not a blocker for this PR but would be a nice experience to keep the grouping upon paste.

![nov-29-2018 13-53-21](https://user-images.githubusercontent.com/25181823/49244997-25aab080-f3df-11e8-927e-aaa9c09689f9.gif)

rashidkpc commented:
I'm personally ok with the whole images not pasting between workpads thing for now, it's something we could maybe do in a phase 2.

cqliu1 commented:
@alexfrancoeur I added your feedback to the clipboard issue, so we can keep track of the features you suggested to improve the experience later on.
